src/lib.rs

```diff
@@ -358,6 +363,175 @@ fn jpeg_size<R: BufRead + Seek>(reader: &mut R) -> ImageResult<ImageSize> {
     })
 }
 
+fn jxl_size<R: BufRead + Seek>(reader: &mut R) -> ImageResult<ImageSize> {
+    let mut file_header = [0; 16]; // The size is variable, but doesn't exceed 16 bytes
+    let mut header_size = 0;
+
+    reader.seek(SeekFrom::Start(0))?;
+    reader.read_exact(&mut file_header[..2])?;
+
+    if &file_header[..2] == b"\xFF\x0A" {
+        // Raw data: Read header directly
+        header_size = reader.read(&mut file_header[2..])? + 2;
+    } else {
+        // Container format: Read from a single jxlc box or multiple jxlp boxes
+        reader.seek(SeekFrom::Start(12))?;
+
+        loop {
+            let (box_type, box_size) = next_tag(reader)?;
+            let box_start = reader.stream_position()? - 8;
+
+            // If box_size is 1, the real size is stored in the first 8 bytes of content.
+            // If box_size is 0, the box ends at EOF.
+
+            let box_size = match box_size {
+                1 => {
+                    let mut box_size = [0; 8];
+                    reader.read_exact(&mut box_size)?;
+                    u64::from_be_bytes(box_size)
+                },
+                _ => box_size as u64,
+            };
+
+            let box_end = box_start.checked_add(box_size).ok_or(ImageError::CorruptedImage)?;
+            let box_header_size = reader.stream_position()? - box_start;
+
+            if box_size < box_header_size && box_size != 0 {
+                return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, format!("Invalid size for {} box: {}", box_type, box_size)).into());
+            }
+
+            // The jxlc box must contain the complete codestream
+
+            if box_type == "jxlc" {
+                let read_end = match box_size {
+                    0 => 16,
+                    _ => std::cmp::min(box_size - box_header_size, 16) as usize,
+                };
+
+                header_size = reader.read(&mut file_header[..read_end])?;
+                break;
+            }
+
+            // Or it could be stored as part of multiple jxlp boxes
+
+            if box_type == "jxlp" {
+                let mut jxlp_index = [0; 4];
+                reader.read_exact(&mut jxlp_index)?;
+
+                if box_size - box_header_size < 4 && box_size != 0 {
+                    return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, format!("Invalid size for jxlp box: {}", box_size)).into());
+                }
+
+                let read_end = match box_size {
+                    0 => 16 - header_size,
+                    _ => std::cmp::min(box_size - box_header_size - 4, 16 - header_size as u64) as usize + header_size,
+                };
+
+                header_size += reader.read(&mut file_header[header_size..read_end])?;
```

Comment:
There is a potential case where this ends up with an invalid range:

> thread '<unnamed>' panicked at 'slice index starts at 9 but ends at 7', /mnt/c/Users/Amoo/Documents/GitHub/imagesize/src/lib.rs:430:49

Input binary:

> 00 00 00 10 6A 78 6C 70 00 00 89 AD FD C3 00 00
> 00 00 00 11 6A 78 6C 70 00 00 89 AD FD 0A 00 7A
> 00 00 00 00 00 6A 78 6C 70 00 00 70 FF

You can test with `blob_size` and the following string:
```rust
b"\x00\x00\x00\x10\x6A\x78\x6C\x70\x00\x00\x89\xAD\xFD\xC3\x00\x00\x00\x00\x00\x11\x6A\x78\x6C\x70\x00\x00\x89\xAD\xFD\x0A\x00\x7A\x00\x00\x00\x00\x00\x6A\x78\x6C\x70\x00\x00\x70\xFF"
```